### ISSUE TITLE
Request to support Apple Neural Engine for SIMD operations

### ISSUE THREAD

**Issue by gmpassos** [dart-lang/sdk#47111]
https://github.com/dart-lang/sdk/blob/307bc3ef2c83541d7145567847c927ffe5f54020/runtime/vm/compiler/backend/il_arm64.cc#L4198

The ARM64 version of Dart, which is already in the beta channel, can easily bring support to Apple Neural Engine for SIMD operations.

After the release of the Neural Engine inside the Apple M1 chip, for computers, the Neural Engine cores can be used through veclib operations (no only from Core ML framework as previously):

https://developer.apple.com/documentation/accelerate/veclib

Any advice on how we can make these changes? I can help to test and fix the branch.

Regards.

**Comment by mraleph** [comment]
Apple Neural Engine does not seem to have any real public documentation, so it is impossible to target it directly. 

For large vector operations you should be able to call `veclib` through FFI. 

Are there any _small_ (128-bit) vector operations you are interested in which are not covered by existing SIMD APIs?

**Comment by gmpassos** [comment]
The "Neural Engine" is the part of the Apple Silicon with cores specialized for ML and vectors.

The way to access them directly, without us of Core ML (framework), is exactly with vecLib as documented at the Apple link above.

**Comment by mraleph** [comment]
> The main idea is to be able to compute a set of SIMD through vecLib, only that. The use of FFI won't match a high performance due overhead.

Ok, let me try to explain: imagine you have a function `f` from `vecLib` there are two possibilities: 

1. it is some sort of intrinsic that C++ compiler can inline and eliminate the call; 
2. it is just a normal C function which can't be inlined.

In the first case (intrinsic) we need to know an expansion of `f` to ARM/ARM64 assembly to be able to transform it into an efficient, no call assembly. 

In the second case (normal function) we need to call the function `f` from Dart code. In this case FFI should in the near future deliver the same performance as generating calls to these functions manually in the backend. (There are some known issues associated with FFI right now that prevent this - but this are going to be fixed). Let me highlight this: the very idea behind FFI is to provide Dart developer with a way to call C functions with the same level of overhead you are going to have when manually emitting these calls in the Dart compiler backend. 

Furthermore, the larger you input data the less FFI/C-call overhead matters.

**Comment by modulovalue** [comment]
> Apple Neural Engine does not seem to have any real public documentation, so it is impossible to target it directly.

AFAIK that's correct. (See: https://github.com/hollance/neural-engine/blob/master/docs/running-on-ane.md) the only way to access ANE is to use higher level Apple approved APIs such as Core ML. ANE will be used if the code has a certain shape, if not, it won't be used. The heuristics are not known and from what I heard, barely anyone outside of Apple is using ANE because performance is unpredictable.

@mraleph I think this can be closed?

**Comment by mraleph** [comment]
Yeah, I don't think we are going to do anything here for the reasons explained above.
